### PR TITLE
feat(doctor): harden Dependencies check against hoisted workspace installs

### LINKED ISSUE #2899: doctor: harden Dependencies check against hoisted/workspace installs
## Context

Identified during review of PR #2896, which fixes the Dependencies preflight check to resolve `node_modules` relative to the server package root (via `import.meta.url`) instead of `process.cwd()`. That was needed because Tauri launches the server with `cwd='/'` under launchd, unconditionally breaking the old `process.cwd()/node_modules` check.

## Description

The new check at `packages/server/src/doctor.js` looks exclusively at `join(SERVER_PKG_DIR, 'node_modules')`. In some workspace/hoisted-install scenarios the deps may live entirely at the workspace root (e.g. `./node_modules`) and the package-local `node_modules/` may be empty or absent.

For this repo's current layout (npm workspaces *with* package-local `node_modules/` populated) the check works. For a monorepo install that hoists all deps, or a `Dockerfile` that uses `npm ci --workspace=@chroxy/server` into a non-standard layout, the check could misreport.

## Proposed approaches

- **createRequire probe:** use `createRequire(import.meta.url).resolve('commander')` (or any required dep) and report pass if it resolves; that matches what Node will actually do at import time.
- **Walk up from SERVER_PKG_DIR** looking for a `node_modules/` dir containing a known dep.
- **Drop the preflight entirely** when running under a bundled `.app` context (tracked separately in #2897): if the app launched, Node already imported the dep graph — any failure would have crashed startup anyway.

## Original comment

> The Dependencies check now only accepts `node_modules` at `join(SERVER_PKG_DIR, 'node_modules')`. In workspace/hoisted installs this directory often doesn't exist even though dependencies are present (e.g. the Docker image installs with `npm ci --workspace=@chroxy/server`, which typically places deps at the workspace root).

— Copilot review on PR #2896

## Acceptance Criteria

- [ ] Dependencies check passes when deps are hoisted to a parent `node_modules/`
- [ ] Dependencies check still catches a genuine "user forgot to run npm install" state
- [ ] Test coverage for both layouts

### PR DESCRIPTION
**Pull request by blamechris**
## Summary

- Doctor's Dependencies preflight checked only `SERVER_PKG_DIR/node_modules` and misreported on npm workspace layouts that hoist deps to a parent `node_modules/` (e.g. `npm ci --workspace=@chroxy/server`, Docker image build).
- New `checkDependencies` helper tries `createRequire(...).resolve(probe)` first (matches Node's real import-time resolution, including hoisted/symlinked trees) and falls back to walking up ancestor directories looking for `node_modules/<probe>/`.
- doctor.js probes three deps (`commander`, `ws`, `@anthropic-ai/claude-agent-sdk`) so a single dep being renamed in the future won't break the preflight.

## Test plan

- [x] Unit test: package-local `node_modules` layout (existing behavior)
- [x] Unit test: hoisted parent `node_modules` layout (new scenario)
- [x] Unit test: neither present — genuine "forgot npm install" fails correctly
- [x] Unit test: multi-probe fallback when one probe is missing
- [x] Unit test: walk-up stops at filesystem root without throwing
- [x] Existing doctor integration tests still pass (13/13)
- [x] Server lint clean — no new warnings

Closes #2899

### CODE REVIEW

**Comment by blamechris** [comment]
## Code Review: PR #2928

### Summary

Fixes Doctor's Dependencies preflight for npm workspace/hoisted layouts (issue #2899). Previously only checked `SERVER_PKG_DIR/node_modules`, which misreports when deps are hoisted to a parent `node_modules/` (e.g. `npm ci --workspace=@chroxy/server`). New `checkDependencies` helper uses a two-tier strategy: `createRequire().resolve()` first (Node's real resolution), then walk-up `node_modules/<probe>/` fallback. Adds multi-probe support so one dep rename won't break preflight.

### Strengths

- **Surgical, well-scoped change** — one new utility, one call site in `doctor.js`, five unit tests. No architectural drift.
- **Dual-strategy resolution** is the right call: `createRequire` handles the common case authoritatively; walk-up fallback survives edge cases (ENOENT races, malformed package.json).
- **Multi-probe design** is forward-looking — resilient to `commander`/`ws`/`@anthropic-ai/claude-agent-sdk` churn in `package.json`.
- **Test coverage** hits all 5 scenarios called out in the acceptance criteria: package-local, hoisted, missing, multi-probe partial, and root-walk termination.
- **Defensive loop cap** (64 levels) and filesystem-root detection (`dirname(dir) === dir`) correctly handle pathological trees.
- **Preserves the Tauri fix** — `SERVER_PKG_DIR` via `import.meta.url` is still the starting point, so `cwd='/'` under launchd is unaffected.

### Issues Found

#### Critical (Must Fix)

None.

#### Suggestions (Should Consider)

| File | Line | Suggestion | Rationale |
|------|------|------------|-----------|
| `packages/server/src/utils/check-dependencies.js` | 54 | Walk-up fallback should verify `<candidate>/package.json` (or similar marker) exists, not just the directory | A bare `node_modules/<probe>/` dir can exist from partial installs, aborted extractions, or stray folders. Returning ok when Node would actually fail at import time defeats the preflight's purpose. Copilot flagged this inline. |
| `packages/server/tests/check-dependencies.test.js` | 30-57 | Fixtures only write `package.json` with no `main`/entry file — `createRequire().resolve()` throws, tests succeed via the walk-up fallback instead of the intended primary path | Tests currently pass coincidentally. Adding a minimal `index.js` and `main: './index.js'` in the fake `package.json`, then asserting `foundAt` ends with `index.js` (not `node_modules/commander`), would actually exercise the `createRequire` branch. Copilot flagged this inline. |
| `packages/server/src/doctor.js` | 99-103 | "createRequire ... matches what Node actually does at import time" is slightly misleading | `createRequire().resolve()` uses CommonJS resolution, which differs from ESM `import` on `exports`/conditions/file extensions. In practice this doesn't matter for the listed probes (they're all packages that resolve fine either way), but the comment oversells the guarantee. Suggest: "matches Node's CommonJS resolution (a reliable proxy for whether deps are installed)". Copilot flagged this inline. |

#### Nitpicks (Optional)

- `check-dependencies.js:51` — the `for (let i = 0; i < 64; i++)` loop will always hit the `dirname(dir) === dir` break first on any real filesystem (max ~40 levels). The cap is pure defense; fine to keep but worth a note that it's belt-and-suspenders.
- `check-dependencies.js:37` — `pathToFileURL(join(startDir, 'package.json'))` works even if `package.json` doesn't exist at `startDir`. That's intentional (it's just a resolution anchor), but a one-line comment noting "file need not exist" would prevent a future reader from adding an unnecessary `existsSync` guard.

### Deferred Items (Follow-Up Issues)

None — all three inline Copilot comments are actionable in this PR and will be handled by the check-pr pass that runs next.

### Architecture Notes

Fits cleanly into the existing preflight pattern in `doctor.js`:
- Mirrors `resolveBinary.js` (sibling in `utils/`) — a focused utility that encapsulates a cross-platform/cross-layout concern with a clean return shape.
- Doesn't touch the WS protocol, Tauri boundary, or session lifecycle — isolated to startup diagnostics.
- `runDoctorChecks()` keeps its flat check array shape, so dashboard/CLI rendering is unaffected.

The multi-probe approach is a nice insurance policy: even if `commander` is ever swapped out, `ws` and `@anthropic-ai/claude-agent-sdk` will continue to anchor the probe — and the Agent SDK package in particular is unlikely to disappear.

### Verdict

**Comment** — Ready to merge once the three Copilot inline suggestions are addressed (walk-up `package.json` marker, test fixtures exercising the `createRequire` path, and doctor.js comment rewording). No blocking issues; the core logic and test scaffolding are sound.

File: packages/server/tests/check-dependencies.test.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    mkdirSync(hoistedDep, { recursive: true })
+    writeFileSync(join(hoistedDep, 'package.json'), JSON.stringify({ name: 'commander', version: '0.0.0' }))
+    mkdirSync(pkgDir, { recursive: true })
+    writeFileSync(join(pkgDir, 'package.json'), JSON.stringify({ name: '@chroxy/server', version: '0.0.0' }))
+
+    const result = checkDependencies({ startDir: pkgDir, probes: ['commander'] })
+    assert.equal(result.ok, true, `expected ok=true, got ${JSON.stringify(result)}`)
+    assert.ok(result.foundAt.includes('commander'))
+    // Must have walked up to the workspace root
+    assert.ok(result.foundAt.startsWith(workspaceRoot), `expected hoisted path under ${workspaceRoot}, got ${result.foundAt}`)
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/server/tests/check-dependencies.test.js]: **FIX**

Fixed in `fbb7d7da2`

**Change:** Test fixtures now include a real `main`/`index.js` so `createRequire().resolve()` succeeds. Assertions verify `foundAt` ends with `commander/index.js` (the resolved file, not the dir) — proving the primary resolution path was exercised rather than the fallback. Also added a new test `falls back to walk-up when createRequire throws but node_modules/<probe>/package.json exists` that intentionally hits only the fallback branch (via a missing `main` file) so both paths now have explicit coverage.

```diff
+ function createResolvableDep(root, name) {
+   const depDir = join(root, 'node_modules', name)
+   mkdirSync(depDir, { recursive: true })
+   writeFileSync(
+     join(depDir, 'package.json'),
+     JSON.stringify({ name, version: '0.0.0', main: 'index.js' }),
+   )
+   writeFileSync(join(depDir, 'index.js'), 'module.exports = {}\n')
+   return depDir
+ }
...
- assert.ok(result.foundAt.includes('commander'))
+ assert.ok(
+   result.foundAt.endsWith(join('commander', 'index.js')),
+   `expected createRequire path to end with commander/index.js, got ${result.foundAt}`,
+ )
```

File: packages/server/src/utils/check-dependencies.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  // 2. Walk-up fallback: look for node_modules/<probe>/ in each ancestor.
+  // Stops at the filesystem root (when dirname returns the same path).
+  let dir = startDir
+  // Safety cap — real trees never exceed a few dozen levels.
+  for (let i = 0; i < 64; i++) {
+    for (const probe of probes) {
+      const candidate = join(dir, 'node_modules', probe)
+      if (existsSync(candidate)) {
+        return { ok: true, foundAt: candidate }
+      }
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/server/src/utils/check-dependencies.js:59]: **FIX**

Fixed in `fbb7d7da2`

**Change:** Walk-up fallback now checks for `<candidate>/package.json` instead of just the directory. This prevents false positives from empty dirs, partially extracted installs, or stray folders that Node would still fail to resolve at import time. Added a regression test `does not false-positive on empty node_modules/<probe>/ dir (no package.json)` to lock this behavior in.

```diff
- // 2. Walk-up fallback: look for node_modules/<probe>/ in each ancestor.
- // Stops at the filesystem root (when dirname returns the same path).
+ // 2. Walk-up fallback: look for node_modules/<probe>/package.json in each
+ // ancestor. Checking package.json (not just the directory) avoids false
+ // positives from empty dirs, partially extracted installs, or stray folders
+ // that Node would still fail to resolve at import time.
+ // Stops at the filesystem root (when dirname returns the same path).
  let dir = startDir
  for (let i = 0; i < 64; i++) {
    for (const probe of probes) {
      const candidate = join(dir, 'node_modules', probe)
-     if (existsSync(candidate)) {
+     if (existsSync(join(candidate, 'package.json'))) {
        return { ok: true, foundAt: candidate }
      }
    }
```

File: packages/server/src/doctor.js [hunk 3]
```diff
@@ -90,15 +91,28 @@ export async function runDoctorChecks({ port, verbose: _verbose } = {}) {
...
   // fail a `${process.cwd()}/node_modules` check.
-  const nodeModulesPath = join(SERVER_PKG_DIR, 'node_modules')
-  if (existsSync(nodeModulesPath)) {
-    checks.push({ name: 'Dependencies', status: 'pass', message: 'node_modules found' })
+  //
+  // Also handles npm workspace hoisting: deps may live in a parent
+  // node_modules/ (e.g. `<repo>/node_modules/commander`) when installed
+  // via `npm ci --workspace=@chroxy/server` at the workspace root. The
+  // helper walks up the tree and uses createRequire to match what Node
+  // actually does at import time.
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/server/src/doctor.js]: **FIX**

Fixed in `fbb7d7da2`

**Change:** Reworded the comment to clarify that `createRequire` uses CommonJS resolution, which is a reliable proxy (not an identical match) for ESM import behavior. For the plain package-name lookups used here (`commander`, `ws`, `@anthropic-ai/claude-agent-sdk`) both resolve the same way, but the comment no longer oversells the guarantee.

```diff
- // Also handles npm workspace hoisting: deps may live in a parent
- // node_modules/ (e.g. `<repo>/node_modules/commander`) when installed
- // via `npm ci --workspace=@chroxy/server` at the workspace root. The
- // helper walks up the tree and uses createRequire to match what Node
- // actually does at import time.
+ // Also handles npm workspace hoisting: deps may live in a parent
+ // node_modules/ (e.g. `<repo>/node_modules/commander`) when installed
+ // via `npm ci --workspace=@chroxy/server` at the workspace root. The
+ // helper walks up the tree and uses createRequire (CommonJS resolution)
+ // as a reliable proxy for whether deps are installed — close enough to
+ // ESM import behavior for plain package-name lookups used here.
```

**Comment by blamechris** [comment]
## Review Comments Addressed

| # | Comment | Outcome | Reference |
|---|---------|---------|-----------|
| 1 | Test fixtures don't exercise the `createRequire` primary path (only `package.json`, no entry file) | FIX | `fbb7d7da2` |
| 2 | Walk-up fallback returns ok for empty `node_modules/<probe>/` dirs — false positive | FIX | `fbb7d7da2` |
| 3 | Comment on `doctor.js:103` oversells `createRequire` as matching ESM import behavior | FIX | `fbb7d7da2` |

**Total:** 3 comments addressed
- Fixed: 3 (commit `fbb7d7da2`)
- False positives: 0
- Follow-up issues created: 0
- Existing issues closed: 0 (issue #2899 will auto-close on merge via `Closes #2899`)

### Test coverage added
- `passes via createRequire when the probe dep is in the package-local node_modules` — now asserts `foundAt` ends with `index.js` (proves primary path exercised, not fallback)
- `passes via createRequire when deps are hoisted to a parent node_modules` — same assertion for hoisted layout
- `falls back to walk-up when createRequire throws but node_modules/<probe>/package.json exists` — explicit coverage of the fallback branch (createRequire fails via missing `main`)
- `does not false-positive on empty node_modules/<probe>/ dir (no package.json)` — regression test for the walk-up false-positive fix

Tests: 7/7 pass. Doctor integration: 13/13 pass. Server lint clean (no new warnings).

**Comment by blamechris** [comment]
Rebased onto main to resolve the conflict with #2973 (provider-aware doctor).

**Analysis:** #2973 added provider-awareness to `runDoctorChecks` but left the Dependencies check using the same naive `existsSync(SERVER_PKG_DIR/node_modules)` that issue #2899 tracks. The hoisted-workspace bug is still present on main — this PR's `checkDependencies` utility + walk-up resolution is additive, not redundant.

**Conflict resolution:** Both imports are now present (`getProvider` from #2973, `checkDependencies` from this PR). The Dependencies check (step 6) now uses `checkDependencies` instead of the bare `existsSync` call; the provider-aware steps 4–5 from #2973 are preserved.

All 7 new unit tests pass. Ready for review.
